characters-program/programs/characters-program/src/state.rs

```diff
@@ -14,34 +16,43 @@ pub struct CharacterAccount {
 
 impl CharacterAccount {
     pub fn new(name: &str, authority: Pubkey, nft_address: Pubkey, owner: Pubkey) -> Result<Self> {
-        require!(
-            name.len() <= NAME_MAX_LENGTH,
-            ErrorCode::MaxNameLengthExceeded
-        );
-
-        let mut new_name = [0u8; NAME_MAX_LENGTH];
-        new_name[..name.len()].copy_from_slice(name.as_ref());
-
-        Ok(Self {
+        let mut account = Self {
             exp: 0,
-            name: new_name,
+            name: [0; NAME_MAX_LENGTH],
             authority,
             owner,
             nft_address,
-        })
+        };
+
+        account.set_name(name)?;
+
+        Ok(account)
     }
 
     pub fn update(&mut self, dto: UpdateCharacterDTO) -> ProgramResult {
         if let Some(name) = dto.name {
-            require!(
-                name.len() <= NAME_MAX_LENGTH,
-                ErrorCode::MaxNameLengthExceeded
-            );
-            self.name[..name.len()].copy_from_slice(name.as_ref());
+            self.set_name(&name)?;
         }
 
-        self.exp += dto.exp.unwrap_or(0);
+        self.set_exp(dto.exp.unwrap_or(0));
 
         Ok(())
     }
+
+    fn set_name(&mut self, name: &str) -> ProgramResult {
+        require!(
+            name.len() <= NAME_MAX_LENGTH,
+            ErrorCode::MaxNameLengthExceeded
+        );
+        let mut new_name = [0u8; NAME_MAX_LENGTH];
+        new_name[..name.len()].copy_from_slice(name.as_ref());
+        self.name = new_name;
+        Ok(())
+    }
+
+    #[inline]
+    /// Increase character experience by `exp`.
+    pub fn set_exp(&mut self, exp: u64) {
```

Comment:
I suggest changing the function name to `add_exp` since it's increasing the exp

```suggestion
    pub fn add_exp(&mut self, exp: u64) {
```

Reply:
But we'll also need to decrease Character's exp. (i.e., when they die?)

Make sure you can decrease the exp in the same function if you send a negative number

Reply:
I was thinking: doesn't it make more sense to give less exp (or no exp at all), instead of decreasing it? Idk, I don't see experience as a status that one'd lose, different from attributes like honor, defense, attack, etc.

Reply:
I think it's cool to die and lose experience, like a traditional RPG. Maybe we should even consider removing honor and working with only experience/level

Reply:
All right! I'm working on that.